tidy3d/components/device/source.py

```diff
@@ -0,0 +1,94 @@
+"""Defines heat material specifications"""
+from __future__ import annotations
+
+from abc import ABC
+from typing import Union, Tuple
+
+import pydantic.v1 as pd
+
+from .viz import plot_params_heat_source
+
+from ..base import cached_property
+from ..base_sim.source import AbstractSource
+from ..data.data_array import TimeDataArray
+from ..viz import PlotParams
+
+from ...constants import VOLUMETRIC_HEAT_RATE
+from ...exceptions import SetupError
+from ...log import log
+
+
+class DeviceSource(AbstractSource, ABC):
+    """Abstract device source."""
+
+    structures: Tuple[str, ...] = pd.Field(
+        title="Target Structures",
+        description="Names of structures where to apply heat source.",
+    )
+
+    @cached_property
+    def plot_params(self) -> PlotParams:
+        """Default parameters for plotting a Source object."""
+        return plot_params_heat_source
+
+    @pd.validator("structures", always=True)
+    def check_non_empty_structures(cls, val):
+        """Error if source doesn't point at any structures."""
+        if len(val) == 0:
+            raise SetupError("List of structures for heat source is empty.")
+
+        return val
+
+
+class HeatSource(DeviceSource):
+    """Volumetric heat source.
+
+    Example
+    -------
+    >>> heat_source = HeatSource(rate=1, structures=["box"])
+    >>> # analytical expression as heat source
+    >>> expression = f"{r}*(sqrt(x) + sin(y)*pow(z,3))"
+    >>> heat_source_expr = HeatSource(rate=expression, structures=["box"])
+    """
+
+    rate: Union[str, float, TimeDataArray] = pd.Field(
```

Comment:
how is the expression validated? I think some explanation of how to do this should be given.

Reply:
Could be very relevant for defining post-processing functions in eg. adjoint so I'm curious how this ends up turning out.

Reply:
We haven't decided yet whether we want to keep the expression here or not. I included this because the CFD guys have an expression evaluator in their [code](https://github.com/flexcompute/Flow360Core/tree/master/Libraries/ExpressionEvaluator) and it was trivial to include it. The issue is that we aren't validating this in the front-/back-end so we don't know if it'll work until we run it, which isn't great. 
I guess one way to check the expression is valid would be to compile a program based on that library that checks the string and have this program called from the validator. This isn't so trivial, so I guess, at this time, it'd be best to remove this option entirely...

Reply:
Maybe let's remove it. but I think we'll start working on this function serialization problem a bit more seriously in the coming months so maybe we can keep this as one of the main applications of it.

Reply:
I actually ran into this issue when working on one of my examples previously, think I fixed it locally with the intention of PRing, but `TimeDataArray` it's not actually supported currently. I think it should be removed or construct multiple individual quasi-static sims.